architecture.md

```diff
@@ -0,0 +1,72 @@
+
+This document describes the high-level architecture of Metals following the philosophy of [rust-analyzer's ARCHITECTURE.md](https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html)
+
+## Other resources
+
+- For the general contributing guidelines, see [Getting started guide for contributors](https://scalameta.org/metals/docs/contributors/getting-started/).
+- If you're interested in the overview of how Metals works, [A Dive into how Metals works](https://youtu.be/fpzN_vTBy18) video will be helpful.
+- For LSP specification, please refer [the LSP document](https://microsoft.github.io/language-server-protocol/), you don't have to read all of the `specification`, just read the `overview` and pick out the interesting parts when you need it.
+
+## LSP endpoints
+
+The most important file of this project is `MetalsLanguageServer.scala`. In this file, we define all of the LSP endpoints, for example, you will find the endpoint for `textDocument/completion` request as follows.
+
+```scala
+@JsonRequest("textDocument/completion")
+def completion(...) = ...
+```
+
+In addition to the LSP endpoints, this file contains many components, for example `private var semanticdbs: Semanticdbs = _`. Those components will be initialized later at some points like on LSP's `initialize` request.
+
+## Presentation compiler
+
+Metals features are powered by presentation compilers, if you hit `Compilers.scala` it is the client of compilers.
+
+`mtags` module is the Scala version specific module used to interact with the Scala presentation compiler. You can find the Java interfaces under `mtags-interface` project.
+
+For example, `ScalaPresentationCompiler.java` in `mtags-interface` is the interface for `ScalaPresentationCompiler.scala` under `scala-2` and `scala-3` directories of `mtags` module.
+
+For more details
+
+- [An introduction to the Scala presentation compiler](https://www.chris-kipp.io/blog/an-intro-to-the-scala-presentation-compiler)
+- Presentation compilers' endpoint
+  - Scala3 `InteractiveDriver.scala` in [lampepfl/dotty](https://github.com/lampepfl/dotty)
+  - Scala2: `interactive/Global.scala` in [scala/scala](https://github.com/scala/scala)
+
+## SemanticDB
+
+Metals uses [semanticdb](https://scalameta.org/docs/semanticdb/guide.html) for many features (Providers that receive the instance of `Semanticdbs`). Metals consumes SemanticDBs in two ways:
+
+- `FileSystemsSemanticdbs` consumes SemanticDBs on disks
+  - Semanticdb will be generated by the build servers to disk, see https://scalameta.org/metals/docs/integrations/new-build-tool#enable-semanticdb
+- `InteractiveSemanticdbs` will generate SemanticDB on the fly using the presentation compiler via `mtags`, in case `FileSystemsSemanticdbs` failed or when the build tool doesn't compiler specific files such as dependency sources or sbt files.
+
+In addition, classes extend `SemanticdbFeatureProvider` index the semanticdb symbols, that will be updated by `SemanticdbIndexer`.
+
+## Build Server
```

Comment:
I'm not sure this is enough, but `architecture.md` is basically for pointing out the endpoint of the topic, so I won't go further and let readers navigate around source code.